### DISCUSSION TITLE
Custom validation message / http response for `MapRequestPayload`

### DISCUSSION THREAD

**Discussion by herndlm** [symfony/symfony#58295] [Q&A] [answered]
Hi, I'm new to this attribute and also not a pro when it comes to framework specific details, was mostly only using dedicated components for now.

Anyway, we're currently looking into migrating a REST API from a combination of FOS Rest Bundle + Sensio Framework Extra using `ParamConverter` to Symfony's core `MapEntity` and `MapRequestPayload`.

We have a custom error response format that adds additional information about the fields that did not validate by having some helper function in controllers like `$this->validate($dto)`.

With `MapRequestPayload` the validation happens already earlier, so I suppose I was wondering if
- it is possible to customize the error message somehow? (I dug into `RequestPayloadValueResolver` already and it looks it is not. or is it recommended to use kernel events for smth like that?)
- validation can be disabled / postponed so that we can call our own logic?

(forgot to tick the "I have research for similar.." checkbox. I did do a bit of a search globally and here :))

**Comment by stof** [comment, marked as answer]
You could use a listener for the ExceptionEvent.

Failures in the resolution of arguments with `MapRequestPayload` will report an `HttpException` with a status code 422 (unless your `MapRequestPayload` explicit changes the status code used for validation failure) with a `ValidationFailedException` (which allows accessing the violation list) as its previous exception. This can allow you to create a custom response for those error.

**Reply by herndlm** [reply]
Lightning fast! Thx stof, I had that idea as well but wasn't sure and didn't want to add too much bias. Will try that, definitely better than my workaround.

**Comment by thedomeffm** [comment]
Because I was also looking for the same customization. 

A solution could look like this:

```php
<?php

// ...

class ValidationExceptionSubscriber implements EventSubscriberInterface
{
    public function __construct(
        private readonly SerializerInterface $serializer,
    ) {}

    public static function getSubscribedEvents(): array
    {
        return [
            ExceptionEvent::class => 'onExceptionEvent',
        ];
    }

    public function onExceptionEvent(ExceptionEvent $event): void
    {
        $exception = $event->getThrowable();
        if (!$exception instanceof UnprocessableEntityHttpException) {
            return;
        }

        $validationException = $exception->getPrevious();
        if (!$validationException instanceof ValidationFailedException) {
            return;
        }

        $event->setResponse(
            new JsonResponse(
                data: ['violations' => $this->serializer->toArray($validationException->getViolations())],
                status: $exception->getStatusCode(),
            ),
        );
    }
}
```
